### ISSUE TITLE
[BUG] `Dataset` serialization silently strips `requires_grad` from PennyLane tensor parameters

### ISSUE THREAD

**Issue by Timeless-hep** [PennyLaneAI/pennylane#9609]
### Expected behavior

After storing a PennyLane operator with trainable parameters (`requires_grad=True`) into a `qml.data.Dataset` and loading it back, the parameters should preserve `requires_grad=True` so that gradient computation continues to work.

### Actual behavior

The loaded operator's parameters come back as plain numpy arrays with `requires_grad=False`. The `requires_grad` metadata is silently discarded during HDF5 serialization. Any subsequent `qml.grad` call on the loaded parameters silently produces incorrect gradients or fails, because the autograd system no longer tracks them as trainable.

This happens because `DatasetPyTree.value_to_hdf5` (`pennylane/data/attributes/pytree.py` lines 45–56) calls `flatten()` to extract leaf tensors, then stores them via `DatasetArray` which converts to numpy. Numpy has no concept of `requires_grad`, so this metadata is lost. The pytree `treedef` only captures tree structure, not per-leaf metadata. On deserialization, values are reconstructed as plain arrays with `requires_grad=False` by default.

### Additional information

Reproduces 100% of the time.

### Source code

```shell
import pennylane as qml
from pennylane import numpy as pnp
import os

param = pnp.tensor(0.5, requires_grad=True)
print(f"Before: requires_grad={param.requires_grad}")  # True

op = qml.RX(param, wires=0)
dataset = qml.data.Dataset(op=op)

filepath = "/tmp/test_grad_roundtrip.h5"
dataset.write(filepath)

loaded = qml.data.Dataset()
loaded.read(filepath)

loaded_param = loaded.op.data[0]
print(f"After:  requires_grad={getattr(loaded_param, 'requires_grad', False)}")  # False — metadata lost

os.remove(filepath)
```

### Tracebacks

```shell
Before: requires_grad=True
After:  requires_grad=False
```

### System information

```shell
Version: 0.45.0
Platform info: macOS
Python version: 3.13.13
```

### Existing GitHub issues

- [x] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by CatalinaAlbornoz** [comment]
Thanks for reporting this bug @YujinSong-hep !
